### PR TITLE
Return Completed ReadResult after stream returns 0 bytes

### PR DESCRIPTION
**Pull request by jkotalik**
For https://github.com/aspnet/AspNetCore/issues/7329. This is a combination of 2 bugs:
1. StreamPipeReader.ReadAsync shouldn't call Stream.ReadAsync if the stream previously return 0 bytes. This PR solves this issue. 
Note: most implementation of Stream continue to return 0 bytes from ReadAsync after the stream is completed.
2. ResponseStream.ReadAsync should probably not complete the PipeReader https://github.com/aspnet/AspNetCore/blob/master/src/Hosting/TestHost/src/ResponseStream.cs#L118. We don't do this in ReadOnlyPipeStream today. This is a problem because calling ReadAsync after it return 0 bytes will throw an exception rather than returning 0 bytes.

### CODE REVIEW

File: src/Http/Http/src/StreamPipeReader.cs [hunk 1]
```diff
@@ -208,6 +208,11 @@ public override void OnWriterCompleted(Action<Exception, object> callback, objec
         public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
         {
             // TODO ReadyAsync needs to throw if there are overlapping reads.
+            if (_isWriterCompleted)
```
Comment (davidfowl) [inline src/Http/Http/src/StreamPipeReader.cs]: This bool name is wrong though. Is _readingCompleted.

File: src/Http/Http/test/StreamPipeReaderTests.cs [hunk 2]
```diff
@@ -661,5 +672,28 @@ public override async ValueTask<int> ReadAsync(Memory<byte> buffer, Cancellation
             }
 #endif
         }
+
+        private class ThrowAfterZeroByteReadStream : MemoryStream
+        {
+            private bool throwOnNextCallToRead;
```
Comment (JamesNK) [inline src/Http/Http/test/StreamPipeReaderTests.cs]: nit: underscore

**Review by JamesNK — APPROVED** [review]

**Review by davidfowl — APPROVED** [review]

File: src/Http/Http/test/StreamPipeReaderTests.cs [hunk 3]
```diff
@@ -599,6 +599,17 @@ public async Task ReadAsyncWithEmptyDataCompletesStream()
...
         }
 
+        [Fact]
+        public async Task ReadAsyncAfterReceivingCompletedReadResultDoesNotThrow()
+        {
+            Stream = new ThrowAfterZeroByteReadStream();
+            Reader = new StreamPipeReader(Stream);
+            var readResult = await Reader.ReadAsync();
+
+            readResult = await Reader.ReadAsync();
```
Comment (Tratcher) [inline src/Http/Http/test/StreamPipeReaderTests.cs]: Assert zero bytes read.

**Review by Tratcher — APPROVED** [review]

**Review by davidfowl — APPROVED** [review]

**Comment by davidfowl** [comment]
@jkotalik played with things a bit and found this https://github.com/aspnet/AspNetCore/issues/7448

**Pull request merged**